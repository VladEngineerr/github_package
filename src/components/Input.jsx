import PropTypes from 'prop-types';
import { forwardRef } from 'react';
import classnames from 'classnames';

const Input = forwardRef(function (
  {
    className,
    disabled = false,
    defaultValue,
    error = false,
    errorMessage,
    label,
    name,
    onBlur = () => {},
    onChange = () => {},
    onFocus = () => {},
    onInput = () => {},
    pattern,
    placeholder,
    required = false,
    testID = 'xInput',
    type = 'text',
  },
  ref
) {
  return (
    <fieldset
      className={classnames('component classes here', className)}
      data-testid={testID}
    >
      <label
        htmlFor={name}
        className={classnames('label classes here', error)}
        data-testid={`${testID}-label`}
      >
        {label}
      </label>
      <input
        ref={ref}
        data-testid={`${testID}-input`}
        className={classnames('input classes here', error)}
        defaultValue={defaultValue}
        onBlur={onBlur}
        onChange={onChange}
        onFocus={onFocus}
        onInput={onInput}
        placeholder={label || placeholder}
        type={type}
        pattern={pattern}
        required={required}
        disabled={disabled}
        name={name}
      />
      {error && (
        <p data-testid={`${testID}-error`} className="error classes here">
          {errorMessage}
        </p>
      )}
    </fieldset>
  );
});

Input.propTypes = {
  className: PropTypes.string, // accept className for placement
  defaultValue: PropTypes.any,
  design: PropTypes.oneOf(['primary', 'secondary', 'link']), // built-in styles
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onInput: PropTypes.func,
  pattern: PropTypes.any,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  testID: PropTypes.string,
  type: PropTypes.oneOf([
    'date',
    'datetime',
    'email',
    'number',
    'password',
    'tel',
    'text',
  ]), // only allowable types, button to ignore form submission
};

export default Input;

