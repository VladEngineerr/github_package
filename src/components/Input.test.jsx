import Input from './Input';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

describe('Input', () => {
  describe('snapshots', () => {
    test('render with defaults', () => {
      const { baseElement } = render(
        <Input label="Test" name="test" onInput={vi.fn} />
      );

      expect(baseElement).toMatchSnapshot();
    });

    test('render with defaultValue', () => {
      const { baseElement } = render(
        <Input
          label="Test"
          name="test"
          defaultValue="foo"
          onInput={vi.fn}
        />
      );

      expect(baseElement).toMatchSnapshot();
    });

    test('render with props', () => {
      const { baseElement } = render(
        <Input
          className="test class"
          label="Test"
          name="test"
          type="number"
          onInput={vi.fn}
          onFunc={vi.fn}
          onBlur={vi.fn}
          placeholder="this is a placeholder"
        />
      );

      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('integration', () => {
    let user;

    beforeEach(() => {
      user = userEvent.setup();
    });

    afterEach(() => {
      user = null;
    });

    test('event handling', async () => {
      const onBlur = vi.fn();
      const onChange = vi.fn();
      const onFocus = vi.fn();
      const onInput = vi.fn();

      render(
        <Input
          label="Test"
          name="test"
          onBlur={onBlur}
          onChange={onChange}
          onFocus={onFocus}
          onInput={onInput}
        />
      );

      const input = screen.getByTestId('xInput-input');

      await user.type(input, 'foobar'); // enter into the input and type

      expect(input).toHaveValue('foobar'); // updated while typing
      expect(onFocus).toHaveBeenCalled(); // fires when entering
      expect(onInput).toHaveBeenCalled(); // fires while typing

      await user.tab(input); // leave the input

      expect(onBlur).toHaveBeenCalled(); // fires when exiting
      expect(onChange).toHaveBeenCalled(); // fires when exiting and value is changed
    });
  });
});
