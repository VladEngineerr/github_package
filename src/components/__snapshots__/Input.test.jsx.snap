// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > snapshots > render with defaultValue 1`] = `
<body>
  <div>
    <fieldset
      class="component classes here"
      data-testid="xInput"
    >
      <label
        class="label classes here"
        data-testid="xInput-label"
        for="test"
      >
        Test
      </label>
      <input
        class="input classes here"
        data-testid="xInput-input"
        name="test"
        placeholder="Test"
        type="text"
        value="foo"
      />
    </fieldset>
  </div>
</body>
`;

exports[`Input > snapshots > render with defaults 1`] = `
<body>
  <div>
    <fieldset
      class="component classes here"
      data-testid="xInput"
    >
      <label
        class="label classes here"
        data-testid="xInput-label"
        for="test"
      >
        Test
      </label>
      <input
        class="input classes here"
        data-testid="xInput-input"
        name="test"
        placeholder="Test"
        type="text"
        value=""
      />
    </fieldset>
  </div>
</body>
`;

exports[`Input > snapshots > render with props 1`] = `
<body>
  <div>
    <fieldset
      class="component classes here test class"
      data-testid="xInput"
    >
      <label
        class="label classes here"
        data-testid="xInput-label"
        for="test"
      >
        Test
      </label>
      <input
        class="input classes here"
        data-testid="xInput-input"
        name="test"
        placeholder="Test"
        type="number"
        value=""
      />
    </fieldset>
  </div>
</body>
`;
